### PR TITLE
More flush reduction for Netty

### PR DESCRIPTION
**Pull request by louiscryan**
@ejona86 @nmittler 

Update to handle flush reduction change in Netty
Check if we're in the Netty event loop, i.e. a read has occurred and we're using direct executor, when writing so flushes can coalesce

This PR is a work in progress but probably worth discussing now

Partially fixes 
https://github.com/grpc/grpc-java/issues/332
and
https://github.com/grpc/grpc-java/issues/331

### CODE REVIEW

**Comment by nicolasnoble** [comment]
Can one of the admins verify this patch?

File: netty/src/main/java/io/grpc/transport/netty/NettyClientStream.java [hunk 1]
```diff
@@ -124,7 +133,7 @@ protected void sendFrame(WritableBuffer frame, boolean endOfStream, boolean flus
       bytebuf = ((NettyWritableBuffer) frame).bytebuf();
     }
     channel.write(new SendGrpcFrameCommand(this, bytebuf, endOfStream));
-    if (flush) {
+    if (flush && !channel.eventLoop().inEventLoop()) {
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientStream.java]: I'm uncertain, but we may possibly be in the same event loop, but for a different Channel. Like if we are servicing a request and make an outgoing call.

File: integration-testing/src/main/java/io/grpc/testing/integration/AbstractTransportTest.java [hunk 2]
```diff
@@ -89,7 +89,7 @@
       ProtoUtils.keyForProto(Messages.SimpleContext.getDefaultInstance());
   private static ScheduledExecutorService testServiceExecutor;
   private static ServerImpl server;
-  private static int OPERATION_TIMEOUT = 5000;
+  private static int OPERATION_TIMEOUT = 5000000;
```
Comment (nmittler) [inline integration-testing/src/main/java/io/grpc/testing/integration/AbstractTransportTest.java]: Can you revert the changes to this file?

File: build.gradle [hunk 3]
```diff
@@ -51,7 +51,8 @@ subprojects {
 
                 // TODO: Unreleased dependencies.
                 // These must already be installed in the local maven repository.
-                netty: 'io.netty:netty-codec-http2:4.1.0.Beta5-SNAPSHOT',
+                //netty: 'io.netty:netty-codec-http2:4.1.0.Beta5-SNAPSHOT',
+                netty: 'io.netty:netty-codec-http2:5.0.0.Alpha3-SNAPSHOT',
```
Comment (nmittler) [inline build.gradle]: Why did you need to switch over to Netty 5?

File: core/src/main/java/io/grpc/transport/MessageDeframer.java [hunk 4]
```diff
@@ -216,49 +217,59 @@ private void checkNotClosed() {
...
-          processBody();
-
-          // Since we've delivered a message, decrement the number of pending
-          // deliveries remaining.
-          pendingDeliveries--;
-          break;
-        default:
-          throw new AssertionError("Invalid state: " + state);
-      }
+    // We can have reentrancy here when using a direct executor, triggered by calls to
```
Comment (nmittler) [inline core/src/main/java/io/grpc/transport/MessageDeframer.java]: Didn't you already submit this change in another PR?  If so, could you rebase?

**Comment by louiscryan** [comment]
PTAL - This is a pretty substantial update

File: integration-testing/src/main/java/io/grpc/testing/integration/AbstractTransportTest.java (hunk 2 above)
Comment (nmittler) [inline integration-testing/src/main/java/io/grpc/testing/integration/AbstractTransportTest.java]: @louiscryan ... same question ... can you revert?

File: integration-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java [hunk 5]
```diff
@@ -293,7 +294,14 @@ private void scheduleNextChunk() {
         Chunk nextChunk = chunks.peek();
         if (nextChunk != null) {
           scheduled = true;
-          executor.schedule(dispatchTask, nextChunk.delayMicroseconds, TimeUnit.MICROSECONDS);
+          if (executor instanceof ScheduledExecutorService) {
```
Comment (nmittler) [inline integration-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java]: why make this change?

File: integration-testing/src/test/java/io/grpc/testing/integration/NettyTransportIoEfficiencyTest.java [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+
+  public static class ClientCountingChannel extends NioSocketChannel {
+
+    public static volatile int flushCount = 0;
+
+    @Override
+    protected void doWrite(ChannelOutboundBuffer in) throws Exception {
+      flushCount++;
+      super.doWrite(in);
+      Thread.dumpStack();
```
Comment (nmittler) [inline integration-testing/src/test/java/io/grpc/testing/integration/NettyTransportIoEfficiencyTest.java]: Do you want to keep this?  Should it be at least commented out for unit tests?

File: netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java [hunk 7]
```diff
@@ -137,15 +137,16 @@ public void channelActive(ChannelHandlerContext ctx) throws Exception {
   @Override
   public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
     try {
-      if (msg instanceof CreateStreamCommand) {
-        createStream((CreateStreamCommand) msg, promise);
+      if (msg instanceof CreateStreamNettyCommand) {
```
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java]: Why did you create new classes for these commands? Couldn't we have just changed the existing classes to have a `flush` boolean?

File: netty/src/main/java/io/grpc/transport/netty/NettyClientTransport.java [hunk 8]
```diff
@@ -167,14 +167,10 @@ public ClientStream newStream(MethodDescriptor<?, ?> method, Metadata.Headers he
...
-    CreateStreamCommand msg = new CreateStreamCommand(http2Headers, stream);
-    if (method.getType().clientSendsOneMessage()) {
-      // No need to flush yet if the client is required to send at least one message
-      channel.write(msg);
-    } else {
-      channel.writeAndFlush(msg);
-    }
-
+    CreateStreamNettyCommand msg = new CreateStreamNettyCommand(
+        Utils.shouldFlush(channel, !method.getType().clientSendsOneMessage()),
```
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientTransport.java]: Can you restore the comment WRT flushing?

File: netty/src/main/java/io/grpc/transport/netty/NettyServer.java [hunk 9]
```diff
@@ -101,6 +101,7 @@ public void start(ServerListener serverListener) throws IOException {
     if (NioServerSocketChannel.class.isAssignableFrom(channelType)) {
       b.option(SO_BACKLOG, 128);
       b.childOption(SO_KEEPALIVE, true);
+      // b.childOption(ChannelOption.TCP_NODELAY, false);
```
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/NettyServer.java]: Was this intentional?

File: netty/src/main/java/io/grpc/transport/netty/NettyServerStream.java [hunk 10]
```diff
@@ -70,12 +70,20 @@ void inboundDataReceived(ByteBuf frame, boolean endOfStream) {
...
   @Override
   public void request(final int numMessages) {
-    channel.eventLoop().execute(new Runnable() {
-      @Override
-      public void run() {
-        requestMessagesFromDeframer(numMessages);
-      }
-    });
+    if (channel.eventLoop().inEventLoop()) {
+      // Processing data read in the event loop so can call into the deframer immediately.
```
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/NettyServerStream.java]: Maybe add a comment as to why we don't need a flush in this case?

File: netty/src/main/java/io/grpc/transport/netty/NettyServerStream.java [hunk 11]
```diff
@@ -85,8 +93,12 @@ protected void inboundDeliveryPaused() {
 
   @Override
   protected void internalSendHeaders(Metadata.Headers headers) {
-    channel.writeAndFlush(new SendResponseHeadersCommand(id(),
-        Utils.convertServerHeaders(headers), false));
+    SendResponseHeadersCommand headersCommand =
+        new SendResponseHeadersCommand(
+            Utils.shouldFlush(channel, !getMethodType().serverSendsOneMessage()),
```
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/NettyServerStream.java]: Add a comment about how the flushing works here.

File: netty/src/main/java/io/grpc/transport/netty/AbstractNettyCommand.java [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+ */
+
+package io.grpc.transport.netty;
+
+import io.netty.channel.ChannelHandlerContext;
+
+/**
+ * Common base class for commands executed on the Netty event loop.
+ */
+public class AbstractNettyCommand {
```
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/AbstractNettyCommand.java]: Is there some way we can combine all of the commands.  It seems like we now use different ones for client vs server. It would be nice to just have a single set.

File: netty/src/main/java/io/grpc/transport/netty/AbstractNettyCommand.java [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.transport.netty;
+
+import io.netty.channel.ChannelHandlerContext;
+
+/**
+ * Common base class for commands executed on the Netty event loop.
```
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/AbstractNettyCommand.java]: Is the distinction here that these are `write` commands as opposed to `read` commands (which don't have a flush)?

**Comment by nmittler** [comment]
@louiscryan finished a first pass ... just a few comments.

File: integration-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java (hunk 5 above)
Comment (louiscryan) [inline integration-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java]: It allows testing a streaming response producer that does its work on the network thread using direct executor.  See the effect in NettyTransportIoEfficiencyTest

smallServerStreaming vs smallServerStreamingDirectExecutor

File: integration-testing/src/test/java/io/grpc/testing/integration/NettyTransportIoEfficiencyTest.java (hunk 6 above)
Comment (louiscryan) [inline integration-testing/src/test/java/io/grpc/testing/integration/NettyTransportIoEfficiencyTest.java]: Artifact from debugging. Removed.

File: netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java (hunk 7 above)
Comment (louiscryan) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java]: Was just trying to make their naming more consistent

File: netty/src/main/java/io/grpc/transport/netty/NettyClientTransport.java (hunk 8 above)
Comment (louiscryan) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientTransport.java]: Done

File: netty/src/main/java/io/grpc/transport/netty/NettyServer.java (hunk 9 above)
Comment (louiscryan) [inline netty/src/main/java/io/grpc/transport/netty/NettyServer.java]: Nope. Reverted

File: netty/src/main/java/io/grpc/transport/netty/NettyServerStream.java (hunk 10 above)
Comment (louiscryan) [inline netty/src/main/java/io/grpc/transport/netty/NettyServerStream.java]: Done, made equivalent change in NettyClientStream

File: netty/src/main/java/io/grpc/transport/netty/NettyServerStream.java (hunk 11 above)
Comment (louiscryan) [inline netty/src/main/java/io/grpc/transport/netty/NettyServerStream.java]: Made comment about flushing behavior for unary calls. The coalescing of flushes after read is doc'd on the Utils.shouldFlush method

File: netty/src/main/java/io/grpc/transport/netty/AbstractNettyCommand.java (hunk 12 above)
Comment (louiscryan) [inline netty/src/main/java/io/grpc/transport/netty/AbstractNettyCommand.java]: No sure what you mean? The command classes are used by both client and server streams

File: netty/src/main/java/io/grpc/transport/netty/AbstractNettyCommand.java (hunk 13 above)
Comment (louiscryan) [inline netty/src/main/java/io/grpc/transport/netty/AbstractNettyCommand.java]: Yes, happy to make naming changes to better indicate this.

**Comment by louiscryan** [comment]
FYI before and after MaxQpsBenchmakr no's

- We are slower at low channel / stream counts, almost certainly due to delayed flushing and not being saturated.
- As counts go up throughput is significantly better, in particular throughput now scales better with concurrent streams on a single connection to as much as 2x
- Throughput maxes out at 4 connections X 1000 streams per connection and degrades after that but still maintains ~2x margin over old code

BEFORE
Benchmark                                    (channelCount)  (maxConcurrentStreams)   Mode  Cnt      Score      Error  Units
MaxQpsBenchmark.measureUnary:callsPerSecond               1                      10  thrpt   20  17233.925 ±  765.450  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond               1                     100  thrpt   20  17494.865 ± 1069.966  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond               1                    1000  thrpt   20  16806.718 ± 2000.184  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond               2                      10  thrpt   20  26280.909 ± 1800.194  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond               2                     100  thrpt   20  30651.307 ± 2506.293  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond               2                    1000  thrpt   20  30850.247 ± 2760.220  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond               4                      10  thrpt   20  42930.248 ± 1463.644  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond               4                     100  thrpt   20  44527.268 ± 1342.846  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond               4                    1000  thrpt   20  46026.291 ± 2705.484  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond               8                      10  thrpt   20  30151.448 ± 1601.576  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond               8                     100  thrpt   20  35231.488 ±  993.181  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond               8                    1000  thrpt   20  37890.380 ± 4221.883  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond              16                      10  thrpt   20  28447.637 ± 1190.671  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond              16                     100  thrpt   20  30699.913 ±  652.101  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond              16                    1000  thrpt   20  28769.108 ± 3196.432  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond              32                      10  thrpt   20  28440.326 ± 1507.982  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond              32                     100  thrpt   20  29217.955 ± 2022.133  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond              32                    1000  thrpt   20  26925.591 ± 3144.744  ops/s

AFTER--------------------------------------
Benchmark                                    (channelCount)  (maxConcurrentStreams)   Mode  Cnt      Score      Error  Units
MaxQpsBenchmark.measureUnary:callsPerSecond               1                      10  thrpt   20  10733.039 ±  344.622  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond               1                     100  thrpt   20  18730.543 ±  937.766  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond               1                    1000  thrpt   20  39446.294 ± 3000.959  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond               2                      10  thrpt   20  24588.238 ± 1022.536  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond               2                     100  thrpt   20  37055.234 ± 2317.336  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond               2                    1000  thrpt   20  66715.266 ± 4290.143  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond               4                      10  thrpt   20  45346.611 ± 1615.482  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond               4                     100  thrpt   20  63213.694 ± 5365.479  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond               4                    1000  thrpt   20  86546.254 ± 2914.231  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond               8                      10  thrpt   20  74125.868 ± 4187.228  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond               8                     100  thrpt   20  70894.528 ± 2297.936  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond               8                    1000  thrpt   20  58071.587 ± 4204.263  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond              16                      10  thrpt   20  49168.631 ± 2329.883  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond              16                     100  thrpt   20  47208.280 ± 1792.296  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond              16                    1000  thrpt   20  45077.976 ± 4848.871  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond              32                      10  thrpt   20  46096.027 ± 2046.095  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond              32                     100  thrpt   20  43651.986 ± 1615.987  ops/s
MaxQpsBenchmark.measureUnary:callsPerSecond              32                    1000  thrpt   20  38481.980 ± 8374.424  ops/s

File: integration-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java (hunk 5 above)
Comment (nmittler) [inline integration-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java]: Do you think there is any advantage to ever using ScheduledExecutor? The original intent was to add delay but I'm not sure we ever care.  Maybe just get rid of it entirely?

File: netty/src/main/java/io/grpc/transport/netty/AbstractNettyCommand.java (hunk 12 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/AbstractNettyCommand.java]: Yeah after I typed that I realized that the distinction isn't client/server ... it's inbound vs outbound.  Maybe name them accordingly so that it's more clear?

File: netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java (hunk 7 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java]: Ok, but maybe we should pick better names to indicate direction of the command (inbound vs outbound).  See my previous comment https://github.com/grpc/grpc-java/pull/367/files#r29995940

File: netty/src/main/java/io/grpc/transport/netty/AbstractNettyCommand.java (hunk 13 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/AbstractNettyCommand.java]: Yeah I think that would be helpful.  Not sure what a good name would be? `inbound` vs `outbound`? `event` vs `command`? WDYT?

**Comment by nmittler** [comment]
@louiscryan nice results! :)  Any thoughts on how we can improve performance for the few streams case?

File: build.gradle [hunk 14]
```diff
@@ -89,7 +89,7 @@ subprojects {
 
                 // TODO: Unreleased dependencies.
                 // These must already be installed in the local maven repository.
-                netty: 'io.netty:netty-codec-http2:4.1.0.Beta5-SNAPSHOT',
+                netty: 'io.netty:netty-codec-http2:4.1.0.Beta6-SNAPSHOT',
```
Comment (nmittler) [inline build.gradle]: Note: before we submit this, we'll want to create the 0.5 branch.  We've already switched master to the Netty 4.1.0-beta5 release.

@ejona86 any thoughts?

File: core/src/main/java/io/grpc/transport/ServerStreamListener.java [hunk 15]
```diff
@@ -31,10 +31,17 @@
...
+import io.grpc.MethodType;
 import io.grpc.Status;
 
 /** An observer of server-side stream events. */
 public interface ServerStreamListener extends StreamListener {
+
+  /**
+   * The {@link MethodType} for the call.
+   */
+  MethodType getMethodType();
```
Comment (nmittler) [inline core/src/main/java/io/grpc/transport/ServerStreamListener.java]: Why do we need this in the listener interface? Seems a little odd.

File: netty/src/main/java/io/grpc/transport/netty/NettyClientStream.java [hunk 16]
```diff
@@ -64,12 +64,22 @@
...
-    channel.eventLoop().execute(new Runnable() {
-      @Override
-      public void run() {
-        requestMessagesFromDeframer(numMessages);
-      }
-    });
+    if (channel.eventLoop().inEventLoop()) {
+      // Processing data read in the event loop so can call into the deframer immediately and since
+      // we are in an event triggered by a read all writes caused by this will be flushed on
+      // completion.
```
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientStream.java]: +1 good comment :)

File: netty/src/main/java/io/grpc/transport/netty/NettyClientTransport.java [hunk 17]
```diff
@@ -166,15 +166,12 @@ public ClientStream newStream(MethodDescriptor<?, ?> method, Metadata.Headers he
...
-    CreateStreamCommand msg = new CreateStreamCommand(http2Headers, stream);
-    if (method.getType().clientSendsOneMessage()) {
-      // No need to flush yet if the client is required to send at least one message
-      channel.write(msg);
-    } else {
-      channel.writeAndFlush(msg);
-    }
-
+    // Write the command requesting the creation of the stream. No need to flush if the
+    // client is required to send at least one message.
```
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientTransport.java]: I'm still not sure I understand why we would ever want to flush here. Why does the number of messages matter?

File: netty/src/main/java/io/grpc/transport/netty/NettyServerStream.java [hunk 18]
```diff
@@ -70,12 +70,22 @@ void inboundDataReceived(ByteBuf frame, boolean endOfStream) {
...
   @Override
   public void request(final int numMessages) {
-    channel.eventLoop().execute(new Runnable() {
-      @Override
-      public void run() {
-        requestMessagesFromDeframer(numMessages);
-      }
-    });
+    if (channel.eventLoop().inEventLoop()) {
+      // Processing data read in the event loop so can call into the deframer immediately and since
```
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/NettyServerStream.java]: So much duplication :/ ... we should really revisit our class/interface hierarchy at some point.

File: netty/src/main/java/io/grpc/transport/netty/NettyServerStream.java [hunk 19]
```diff
@@ -85,8 +95,13 @@ protected void inboundDeliveryPaused() {
 
   @Override
   protected void internalSendHeaders(Metadata.Headers headers) {
-    channel.writeAndFlush(new SendResponseHeadersCommand(id(),
-        Utils.convertServerHeaders(headers), false));
+    // No need to flush headers yet if the server is required to respond with one message.
```
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/NettyServerStream.java]: Same comment as the client case ... why would we ever want to flush when we're just sending headers?

**Comment by nmittler** [comment]
@louiscryan just finished another round.  This looks great, and so do the numbers! :)

File: build.gradle (hunk 14 above)
Comment (ejona86) [inline build.gradle]: In discussion with @louiscryan on Friday, it seemed we both agreed that we probably wouldn't want to bumping our version of Netty to a SNAPSHOT for a non-essential change. The localhost fix is nice, but probably not enough reason to swap to a SNAPSHOT.

File: netty/src/main/java/io/grpc/transport/netty/Utils.java [hunk 20]
```diff
@@ -169,6 +170,18 @@ private static Http2Headers convertMetadata(Metadata headers) {
...
+  /**
+   * Should the channel be asked to flush this write. Checks if the calling thread is the event
+   * loop and assumes that a flush is following if that is the case. This is true for all event
+   * loop work following a read as onReadComplete triggers a flush. For other work scheduled
+   * onto the event loop outside of a read the caller (or event) must schedule a flush.
+   */
+  static boolean shouldFlush(Channel channel, boolean isLastWriteInSequence) {
+    // Don't flush is we are in the event loop as we have been triggered by a read event and
+    // a flush will occur on channel read completion.
+    return !channel.eventLoop().inEventLoop() && isLastWriteInSequence;
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/Utils.java]: As I mentioned before, this doesn't seem strict enough. Multiple channels will use the same event loop, so just being on the right event loop doesn't seem enough to know that channelReadComplete() will be called for _this_ channel.

File: build.gradle (hunk 14 above)
Comment (nmittler) [inline build.gradle]: There is actually another change that we need from beta6: https://github.com/netty/netty/commit/dca6cea78ef6f58d65125fcd268710bd86b86c22.  So I think we need to upgrade to the snapshot regardless.  Let's just cut the tag this week so we can progress on master.

Comment (ejona86) [inline build.gradle]: Can't someone who needs netty/netty@dca6cea just depend on it in their application?

Comment (nmittler) [inline build.gradle]: Yes they can do that ... but what's the longer-term plan for absorbing Netty changes?  Eventually we're going to need something from an unreleased version of Netty.  Or are you thinking that we just have to wait until the next beta release to absorb the changes (which can be weeks)?

Comment (ejona86) [inline build.gradle]: I'm thinking for non-required changes we wait until the next beta release. For important changes, then yes, we use unreleased version of Netty and we delay our release until Netty does theirs.

Comment (nmittler) [inline build.gradle]: sgtm

File: integration-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java (hunk 5 above)
Comment (louiscryan) [inline integration-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java]: Allowing for ScheduleExecutor allows us to do tests where the synchronization points are less predictable which is a good thing. The behavior being demonstrated here is probably atypical.

File: netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java (hunk 7 above)
Comment (louiscryan) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java]: All of the commands are 'outbound' in the sense that the cause writes so I'm not sure theres an improvement to naming in that approach.

File: core/src/main/java/io/grpc/transport/ServerStreamListener.java (hunk 15 above)
Comment (louiscryan) [inline core/src/main/java/io/grpc/transport/ServerStreamListener.java]: This is how the server stream knows the difference between unary & non-unary calls for flush optimization. It has to be on the listener as this is only known by the application layer and is resolved after server stream creation.

File: netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java (hunk 7 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java]: Yup, sorry for my confusion .... it's fine as-is.

File: integration-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java (hunk 5 above)
Comment (nmittler) [inline integration-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java]: sgtm

File: core/src/main/java/io/grpc/transport/ServerStreamListener.java (hunk 15 above)
Comment (nmittler) [inline core/src/main/java/io/grpc/transport/ServerStreamListener.java]: ick ... ok :)

File: netty/src/main/java/io/grpc/transport/netty/NettyClientTransport.java (hunk 17 above)
Comment (louiscryan) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientTransport.java]: If its a unary call we delay flushing the headers as one payload must follow immediately, for client-streaming we don't know if any more messages are coming or when they're coming so we need to flush to inform the server that the call has started.

File: netty/src/main/java/io/grpc/transport/netty/NettyServerStream.java (hunk 19 above)
Comment (louiscryan) [inline netty/src/main/java/io/grpc/transport/netty/NettyServerStream.java]: same response as above

File: netty/src/main/java/io/grpc/transport/netty/Utils.java (hunk 20 above)
Comment (louiscryan) [inline netty/src/main/java/io/grpc/transport/netty/Utils.java]: Agreed, we need a better form of closure over a channel read.

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @louiscryan isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
CLA GitHub bot_
